File: spns/hive/snode.py
from __future__ import annotations
from .swarmpubkey import SwarmPubkey

from collections import deque
from oxenc import bt_serialize
import oxenmq
import random
import time
from typing import Optional


# Maximum size of simultaneous subscriptions in a single subscription request; if we overflow then
# any stragglers wait until the next request, delaying them by a few seconds.  (This is not a rock
# hard limit: we estimate slightly and stop as soon as we exceed it, which means we can go over it a
# bit after appending the last record).
SUBS_REQUEST_LIMIT = 5_000_000

# How long (in seconds) after a successful subscription before we re-subscribe; each subscription
# gets a uniform random value between these two values (to spread out the renewal requests a bit).
RESUBSCRIBE_MIN = 45 * 60
RESUBSCRIBE_MAX = 55 * 60

# How long we wait (in seconds) after a connection failure to a snode storage server before
# re-trying the connection; we use the first value after the first failure, the second one after the
# second failure, and so on (if we run off the end we use the last value).
CONNECT_COOLDOWN = [10, 30, 60, 120]


class SNode:
    """Class managing a connection to a single service node"""

    def __init__(
        self,
        logger,
        hivemind: HiveMind,
        omq: oxenmq.OxenMQ,
        addr: oxenmq.Address,
        swarm: Optional[int],
    ):
        self.logger = logger
        self.hivemind = hivemind
        self.omq = omq
        self.addr = addr
        self.swarm = swarm
        self.conn = None
        self.connected = False
        self.subs = set()  # SwarmPubkey
        self.next = deque()  # [(SwarmPubkey, next), ...] sorted by next re-subscription time
        self.cooldown_until = None
        self.cooldown_fails = 0

        self.connect()

    def __del__(self):
        self.disconnect()

    def connect(self, addr: Optional[oxenmq.Address] = None):
        """Initiates a connection, if not already connected.  If addr is given and doesn't match the
        current address then we drop the current connection first.

        Does nothing if already connected, or already establishing a connection.

        After this call `self.conn` will be set to the ConnectionID of the established or
        establishing connection."""
        if addr is not None and addr != self.addr:
            self.logger.debug(f"disconnecting; addr changing from {self.addr} to {addr}")
            self.disconnect()
            self.addr = addr

        if self.conn is None:
            if self.hivemind._allow_connect():
                self.conn = self.omq.connect_remote(
                    self.addr,
                    on_success=self.on_connected,
                    on_failure=self.on_connect_fail,
                    auth_level=oxenmq.AuthLevel.basic,
                )
                self.logger.debug(f"Establishing connection to {self.addr.full_address}")

    def disconnect(self):
        self.logger.debug(f"disconnecting from {self.addr.full_address}")
        self.connected = False
        if self.conn is not None:
            self.omq.disconnect(self.conn)
            self.conn = None

    def on_connected(self, conn: oxenmq.ConnectionID):
        with self.hivemind.lock:
            self.logger.debug(f"Connected established to {self.addr.full_address}")
            self.cooldown_fails = 0
            self.cooldown_until = None
            self.hivemind._finished_connect()
            if self.conn is None:
                # Our conn got replaced from under us, which probably means we are disconnecting, so do
                # nothing.
                pass

            self.connected = True
            # We either just connected or reconnected, so reset any re-subscription times (so that after
            # a reconnection we force a re-subscription for everyone):
            for n in self.next:
                n[1] = 0.0
            self.check_subs(initial_subs=True)

    def on_connect_fail(self, conn: oxenmq.ConnectionID, reason: str):
        with self.hivemind.lock:
            self.hivemind._finished_connect()
            cooldown = CONNECT_COOLDOWN[
                -1 if self.cooldown_fails >= len(CONNECT_COOLDOWN) else self.cooldown_fails
            ]
            self.cooldown_fails += 1
            self.cooldown_until = time.time() + cooldown
            self.logger.warning(
                f"Connection to {self.addr.full_address} failed: {reason} ({self.cooldown_fails} consecutive failure(s); retrying in {cooldown}s)"
            )
            self.disconnect()

    def add_account(self, account: SwarmPubkey, force_now: bool = False):
        """Adds a new account to be signed up for subscriptions, if it is not already subscribed.
        The new account's subscription will be submitted to the SS the next time check_subs() is
        called (either automatically or manually).

        If `force_now` is True then the account is scheduled for subscription at the next update
        even if already exists."""
        if account not in self.subs:
            self.subs.add(account)
            self.next.appendleft((account, 0.0))
        elif force_now:
            for n in self.next:
                if n[0] == account:
                    n[0] = None  # lazy deletion; we'll skip this when draining the queue
                    break
            self.next.appendleft((account, 0.0))

    def reset_swarm(self, swarm: int):
        """Called when swarm changes; all current subscriptions are dropped."""
        self.next.clear()
        self.subs.clear()
        self.swarm = swarm

    def recheck_swarm_members(self):
        """Called when the network swarm list has changed (but this swarm hasn't), to eject any
        swarm subscriptions that don't belong here anymore.  The Accounts in `.subs` should have had
        .update_swarm called already to check/reset the new swarm_id.  Any existing subscribers that
        are no longer in this swarm will be removed.  (Even without a swarm change of this node,
        this can happen if another new swarm is created next to us).

        This isn't responsible for adding *new* swarm members: this is just called as a first step
        for removing any that shouldn't be here anymore."""
        for n in self.next:
            if n[0].swarm != self.swarm:
                self.subs.remove(n[0])
                n[0] = None  # lazy deletion

    def check_subs(self, *, initial_subs=False):
        """
        Check our subscriptions to resubscribe to any that need it.  If initial is True then this is
        the initial request and we fire off a batch of subscriptions and then another batch upon
        reply, etc. until there are no more subs to send; otherwise we fire off just up to
        SUBS_LIMIT re-subscriptions.
        """
        if not self.connected:
            if self.conn is not None:
                return  # We're already trying to connect

            # If we failed recently we'll be in cooldown mode for a while, so might not connect
            # right away yet.
            if self.cooldown_until is not None:
                if self.cooldown_until > time.time():
                    return
                self.cooldown_until = None

            # We'll get called automatically as soon as the connection gets established, so just
            # make sure we are already connecting and don't do anything else for now.
            return self.connect()

        req = []
        now = time.time()
        req_size = 0
        while req_size < SUBS_REQUEST_LIMIT and self.next and self.next[0][1] <= now:
            acct = self.next.popleft()[0]
            if acct is None:  # lazy deletion; ignore this entry
                continue

            subs = self.hivemind.subscribers.get(acct, now)
            if subs is None:
                continue
            continue

            for sub in subs:
                d = {"n": sub.namespaces, "t": sub.sig_ts, "s": sub.sig}

                account = acct.id
                if account[0] == 0x05:
                    d["P"] = acct.ed25519_pubkey
                else:
                    d["p"] = account

                # bencoded serialized lengths of these data.  Each key is 3 (e.g. "1:n"), list
                # and integer markers add another 2 (l...e or i...e), and binary data (for the sizes
                # we use here) adds 3 (32:, 33: or 64:).
                #
                # The 4 here on the number of namespaces is definitely handwavey: if you were to
                # subscribe to *every* single 65536 namespaces then the encoded namespace list
                # (i-32768ei-32767e...) would be 469304 bytes long, which gives an average of 7.16
                # characters per namespace.  But typically Session is subscribing to something more
                # like 0, 1, 2, 3, ..., up to maybe 6 or 7, all of which are 3 (i0e, etc.).  So 4
                # seems reasonable, and is probably likely to overestimate in most cases.
                req_size += 5 + 4 * len(sub.namespaces) + 15 + 70 + 39

                if sub.subkey_tag:
                    d["S"] = sub.subkey_tag
                    req_size += 38  # 1:S32:... = 6 + the ... (32)

                if sub.want_data:
                    d["d"] = 1
                    req_size += 6  # 1:di1e

                req.append(d)

            self.next.append((acct, now + random.uniform(RESUBSCRIBE_MIN, RESUBSCRIBE_MAX)))

        if req:
            on_reply = None
            if initial and len(req) >= SUBS_REQUEST_LIMIT:
                # We're doing the initial subscriptions, so continue as soon as we get the reply so
                # that we subscribe quickly, but never have more than one massive subscription
                # request active at a time
                on_reply = lambda _: self.connect(initial_subs=True)

            # Otherwise we ignore the reply: even if there are some failures, we have redundancy
            # because we are connected to all swarm members, so even if we lose our subscription for
            # a while from this SN, it shouldn't break notifications.

            self.omq.request(self.conn, "monitor.messages", bt_serialize(req), on_reply=on_reply)
            self.logger.debug(f"(Re-)subscribing to {len(req)} accounts from {self.addr}")

File: spns/hive/snode.cpp
#include "snode.hpp"

#include <oxenc/bt_producer.h>

#include <algorithm>
#include <cassert>
#include <iterator>
#include <memory>
#include <mutex>
#include <oxen/log.hpp>
#include <random>
#include <string>

#include "../bytes.hpp"
#include "../hivemind.hpp"

namespace spns::hive {

namespace log = oxen::log;

static auto cat = log::Cat("snode");

using namespace std::literals;

thread_local std::mt19937_64 rng{std::random_device{}()};

SNode::SNode(HiveMind& hivemind, oxenmq::OxenMQ& omq, oxenmq::address addr, uint64_t swarm) :
        hivemind_{hivemind},
        omq_{omq},
        addr_{std::move(addr)},
        swarm_{swarm}

{
    connect();
}

void SNode::connect() {
    std::lock_guard lock{mutex_};

    if (!conn_) {
        if (hivemind_.allow_connect()) {
            conn_ = omq_.connect_remote(
                    addr_,
                    [this](oxenmq::ConnectionID c) { on_connected(c); },
                    [this](oxenmq::ConnectionID c, std::string_view err) {
                        on_connect_fail(c, err);
                    },
                    oxenmq::AuthLevel::basic);
            log::debug(cat, "Establishing connection to {}", addr_.full_address());
        }
    }
}

void SNode::connect(oxenmq::address addr) {
    bool reconnect;
    {
        std::lock_guard lock{mutex_};
        reconnect = addr != addr_;
    }

    if (reconnect) {
        log::debug(
                cat,
                "disconnecting; addr changing from {} to {}",
                addr_.full_address(),
                addr.full_address());
        disconnect();
        {
            std::lock_guard lock{mutex_};
            addr_ = std::move(addr);
        }
    }

    connect();
}

void SNode::disconnect() {
    std::lock_guard lock{mutex_};

    log::debug(cat, "disconnecting from {}", addr_.full_address());
    connected_ = false;
    if (conn_) {
        omq_.disconnect(conn_);
        conn_ = {};
    }
}

void SNode::on_connected(oxenmq::ConnectionID c) {
    bool no_conn = false;
    {
        std::lock_guard lock{mutex_};

        log::debug(cat, "Connection established to {}", addr_.full_address());
        cooldown_fails_ = 0;
        cooldown_until_.reset();

        if (!conn_) {
            // Our conn got replaced from under us, which probably means we are disconnecting, so do
            // nothing.
            no_conn = true;
        } else {
            // We either just connected or reconnected, so reset any re-subscription times (so that
            // after a reconnection we force a re-subscription for everyone):
            auto now = system_clock::now();
            for (auto& [id, next] : next_)
                next = system_epoch;

            connected_ = true;
        }
    }

    hivemind_.finished_connect();

    if (!no_conn)
        hivemind_.check_my_subs(*this, true);
}

void SNode::on_connect_fail(oxenmq::ConnectionID c, std::string_view reason) {
    {
        std::lock_guard lock{mutex_};

        auto cooldown = cooldown_fails_ >= CONNECT_COOLDOWN.size()
                              ? CONNECT_COOLDOWN.back()
                              : CONNECT_COOLDOWN[cooldown_fails_];
        cooldown_until_ = steady_clock::now() + cooldown;
        cooldown_fails_++;

        log::warning(
                cat,
                "Connection to {} failed: {} ({} consecutive failure(s); retrying in {}s)",
                addr_.full_address(),
                reason,
                cooldown_fails_,
                cooldown.count());

        connected_ = false;
        conn_ = {};
    }

    hivemind_.finished_connect();
}

/// Adds a new account to be signed up for subscriptions, if it is not already subscribed.
/// The new account's subscription will be submitted to the SS the next time check_subs() is
/// called (either automatically or manually).
///
/// If `force_now` is True then the account is scheduled for subscription at the next update
/// even if already exists.
void SNode::add_account(const SwarmPubkey& account, bool force_now) {
    std::lock_guard lock{mutex_};

    auto [it, inserted] = subs_.insert(account);
    if (inserted)
        next_.emplace_front(*it, system_epoch);
    else if (force_now) {
        // We're asked to treat it as "now", so go look for it in the queue and clear it first,
        // then re-insert at the beginning of the queue.
        for (auto& [acc, next] : next_) {
            if (acc && *acc == account) {
                acc.reset();  // lazy deletion; we'll skip this when draining the queue
                break;
            }
        }
        next_.emplace_front(account, system_epoch);
    }
}

void SNode::reset_swarm(uint64_t new_swarm) {
    std::lock_guard lock{mutex_};

    next_.clear();
    subs_.clear();
    swarm_ = new_swarm;
}

void SNode::remove_stale_swarm_members(const std::vector<uint64_t>& swarm_ids) {
    std::lock_guard lock{mutex_};

    for (auto& s : subs_)
        s.update_swarm(swarm_ids);
    for (auto& [acc, next] : next_) {
        if (acc && acc->swarm != swarm_) {
            subs_.erase(*acc);
            acc.reset();
        }
    }
}

void SNode::check_subs(
        const std::unordered_map<SwarmPubkey, std::vector<hive::Subscription>>& all_subs,
        bool initial_subs,
        bool fast) {
    if (!connected_) {
        {
            std::lock_guard lock{mutex_};

            if (conn_)
                return;  // We're already trying to connect

            // If we failed recently we'll be in cooldown mode for a while, so might not connect
            // right away yet.
            if (cooldown_until_) {
                if (*cooldown_until_ > steady_clock::now())
                    return;
                cooldown_until_.reset();
            }
        }

        // We'll get called automatically as soon as the connection gets established, so just
        // make sure we are already connecting and don't do anything else for now.
        return connect();  // NB: must not hold lock when calling this
    }

    std::string req_body = "l";  // We'll add the "e" later
    auto now = system_clock::now();

    size_t next_added = 0, req_count = 0;

    std::lock_guard lock{mutex_};
    while (req_body.size() < SUBS_REQUEST_LIMIT && !next_.empty()) {
        const auto& [maybe_acct, next] = next_.front();
        if (next > now)
            break;
        if (fast && next > system_epoch)
            break;

        if (!maybe_acct) {
            next_.pop_front();
            continue;  // lazy deletion; ignore this entry
        }

        const auto& acct = *maybe_acct;

        auto subs = all_subs.find(acct);
        if (subs == all_subs.end()) {
            next_.pop_front();
            continue;
        }

        std::vector<char> buf;
        for (const auto& sub : subs->second) {

            // Size estimate; this can be over, but mustn't be under the actual size we'll need.
            constexpr size_t base_size = 0 + 3 +
                                         12      // 1:t and i...e where ... is a 10-digit timestamp
                                       + 3 + 67  // 1:s and 64:...
                                       + 3 + 36  // 1:p and 33:... (also covers 1:P and 32:...)
                                       + 3 + 2   // 1:n and the le of the l...e list
                                       + 3 + 3   // 1:d and i1e (only if want_data)
                                       + 3 + 67 + 3 + 39 // 1:S, 64:..., 1:T, 36:... (for subaccount auth)
                    ;

            // The biggest int expression we have is i-32768e; this is almost certainly overkill
            // most of the time though, but no matter.
            auto size = base_size + sub.namespaces.size() * 8;

            auto old_size = req_body.size();
            req_body.resize(old_size + size);

            char* start = req_body.data() + old_size;

            oxenc::bt_dict_producer dict{start, size};

            // keys in ascii-sorted order!
            if (acct.session_ed)
                dict.append("P", acct.ed25519.sv());
            if (sub.subaccount) {
                dict.append("S", sub.subaccount->sig.sv());
                dict.append("T", sub.subaccount->tag.sv());
            }
            if (sub.want_data)
                dict.append("d", 1);
            dict.append_list("n").append(sub.namespaces.begin(), sub.namespaces.end());
            if (!acct.session_ed)
                dict.append("p", acct.id.sv());
            dict.append("s", sub.sig.sv());
            dict.append("t", sub.sig_ts);

            // Resize away any extra buffer space we didn't fill
            req_body.resize(dict.end() - req_body.data());

            req_count++;
        }

        auto delay = 1s * std::uniform_int_distribution<int>{
                                  RESUBSCRIBE_MIN.count(), RESUBSCRIBE_MAX.count()}(rng);

        next_.emplace_back(acct, now + delay);
        next_added++;
        next_.pop_front();
    }

    if (req_body.size() == 1)  // just the initial "l"
        return;

    req_body += 'e';

    // The randomness of our delay will mean the tail of the list isn't sorted, so re-sort from
    // the lowest possible value we could have inserted (now + RESUBSCRIBE_MIN) to the end.

    // Everything we didn't touch should already be sorted:
    assert(std::is_sorted(
            next_.begin(), std::prev(next_.end(), next_added), [](const auto& a, const auto& b) {
                return a.second < b.second;
            }));

    auto it = std::partition_point(
            next_.begin(),
            std::prev(next_.end(), next_added),
            [&now](const decltype(next_)::value_type& n) {
                return n.second < now + RESUBSCRIBE_MIN;
            });

    std::sort(it, next_.end(), [](const auto& a, const auto& b) { return a.second < b.second; });

    // Now everything should be sorted:
    assert(std::is_sorted(next_.begin(), next_.end(), [](const auto& a, const auto& b) {
        return a.second < b.second;
    }));

    auto on_reply = [this, right_away = initial_subs && req_body.size() >= SUBS_REQUEST_LIMIT](
                            bool success, std::vector<std::string> data) {
        if (!success) {
            // TODO: log something about failed request, but otherwise ignore it.  We don't
            // worry about the subscriptions that might lapse because we have full swarm
            // redundancy so it really doesn't matter if a subscription with one or two of
            // the swarm members times out.
        }
        if (right_away) {
            // We're doing the initial subscriptions, and sent a size-limited request so we
            // likely have more that we want to subscribe to ASAP: so we continue as soon as
            // we get the reply back so that we're subscribing as quickly as possible
            // without having more than one (large) subscription request in flight at a
            // time.
            hivemind_.check_my_subs(*this, true);
        }
    };

    omq_.request(conn_, "monitor.messages", std::move(on_reply), std::move(req_body));
    log::debug(cat, "(Re-)subscribing to {} accounts from {}", req_count, addr_.full_address());
}

}  // namespace spns::hive
